`SELECT DISTINCT` SQL fails when useServerPrepStmts=true

After #9114, the IT engine fails to run well with the `SELECT DISTINCT` SQL. It is necessary to find the cause and fix it.

I found a similar problem when PR integration test, e.g. `Integration Test / MySQL-Proxy with DB`, part of error output:
```
Error:  assertExecute[proxy: db -> MySQL -> Placeholder -> SELECT SUM(DISTINCT order_id) FROM t_order WHERE order_id < 1100](org.apache.shardingsphere.test.integration.engine.it.dql.GeneralDQLIT)  Time elapsed: 0.007 s  <<< FAILURE!
java.lang.AssertionError: 
Size of actual result set is different with size of expected dat set rows.
Expected: is <1>
     but: was <0>
```

Steps to reproduce on local environment:

1, Prepare configurations for proxy, copy config files from `shardingsphere-integration-test-suite/src/test/resrouces/docker/db/proxy/conf/` into `shardingsphere-proxy-bootstrap/src/main/resources/`
- copy `config-db.yaml` into `config-sharding.yaml`
- copy `server.yaml` into `server.yaml`

2, Update `shardingsphere-integration-test-suite/src/test/resources/env/engine-env.properties`
```
it.adapters=proxy

#it.scenarios=db,tbl,dbtbl_with_replica_query,replica_query,shadow
it.scenarios=db

#it.databases=H2,MySQL,Oracle,SQLServer,PostgreSQL
it.databases=MySQL
```

Update `shardingsphere-integration-test-suite/src/test/resources/env/db/scenario-env.properties`, `mysql` and `proxy` related configuration, e.g.
```
it.db.mysql.host=127.0.0.1
it.db.mysql.port=3306
it.db.mysql.username=root
it.db.mysql.password=

it.db.proxy.host=127.0.0.1
it.db.proxy.port=3307
```

3, Copy `org.apache.shardingsphere.proxy.Bootstrap.java` to `shardingsphere-integration-test-suite` as `ProxyBootstrapTest.java`, since `STANDARD_TEST` in `config-db.yaml` is a customized algorithm that is implemented in unit test

4, Run `ProxyBootstrapTest.java`

5, Prepare  an test client and start it, e.g. `TempTest.java`
```
public class TempTest {
    
    public static void main(String[] args) throws Exception {
        String sql;
        sql = "SELECT SUM(DISTINCT order_id) s FROM t_order WHERE order_id < 1100"; // result missed in response
//        sql = "SELECT SUM(order_id) s FROM t_order WHERE order_id < 1100"; //ok
//        sql = "SELECT COUNT(DISTINCT order_id) s FROM t_order WHERE order_id < 1100"; //ok
        String jdbcUrl = "jdbc:mysql://localhost:3307/db?useServerPrepStmts=true&serverTimezone=UTC&useSSL=false&useLocalSessionState=true&characterEncoding=utf-8";
        try (Connection connection = DriverManager.getConnection(jdbcUrl, "root", "root");
             PreparedStatement statement = connection.prepareStatement(sql)) {
            try (ResultSet resultSet = statement.executeQuery()) {
                while (resultSet.next()) {
                    System.out.println("c1=" + resultSet.getString(1));
                }
            }
        }
    }
}
```

--- a/shardingsphere-db-protocol/shardingsphere-db-protocol-mysql/src/main/java/org/apache/shardingsphere/db/protocol/mysql/packet/command/query/binary/execute/MySQLBinaryResultSetRowPacket.java
+++ b/shardingsphere-db-protocol/shardingsphere-db-protocol-mysql/src/main/java/org/apache/shardingsphere/db/protocol/mysql/packet/command/query/binary/execute/MySQLBinaryResultSetRowPacket.java
@@ -69,8 +69,9 @@ public final class MySQLBinaryResultSetRowPacket implements MySQLPacket {
     
     private void writeValues(final MySQLPacketPayload payload) {
         for (BinaryCell each : row.getCells()) {
-            if (null != each.getData()) {
-                MySQLBinaryProtocolValueFactory.getBinaryProtocolValue(each.getColumnType()).write(payload, each.getData());
+            Object data = each.getData();
+            if (null != data) {
+                MySQLBinaryProtocolValueFactory.getBinaryProtocolValue(each.getColumnType()).write(payload, data);
             }
         }
     }

--- a/shardingsphere-db-protocol/shardingsphere-db-protocol-mysql/src/main/java/org/apache/shardingsphere/db/protocol/mysql/packet/command/query/binary/execute/protocol/MySQLInt4BinaryProtocolValue.java
+++ b/shardingsphere-db-protocol/shardingsphere-db-protocol-mysql/src/main/java/org/apache/shardingsphere/db/protocol/mysql/packet/command/query/binary/execute/protocol/MySQLInt4BinaryProtocolValue.java
@@ -19,6 +19,8 @@ package org.apache.shardingsphere.db.protocol.mysql.packet.command.query.binary.
 
 import org.apache.shardingsphere.db.protocol.mysql.payload.MySQLPacketPayload;
 
+import java.math.BigDecimal;
+
 /**
  * Binary protocol value for int4 for MySQL.
  */
@@ -31,6 +33,6 @@ public final class MySQLInt4BinaryProtocolValue implements MySQLBinaryProtocolVa
     
     @Override
     public void write(final MySQLPacketPayload payload, final Object value) {
-        payload.writeInt4((Integer) value);
+        payload.writeInt4(value instanceof BigDecimal ? ((BigDecimal) value).intValue() : (Integer) value);
     }
 }

--- a/shardingsphere-db-protocol/shardingsphere-db-protocol-mysql/src/main/java/org/apache/shardingsphere/db/protocol/mysql/packet/command/query/binary/execute/protocol/MySQLInt8BinaryProtocolValue.java
+++ b/shardingsphere-db-protocol/shardingsphere-db-protocol-mysql/src/main/java/org/apache/shardingsphere/db/protocol/mysql/packet/command/query/binary/execute/protocol/MySQLInt8BinaryProtocolValue.java
@@ -33,6 +33,12 @@ public final class MySQLInt8BinaryProtocolValue implements MySQLBinaryProtocolVa
     
     @Override
     public void write(final MySQLPacketPayload payload, final Object value) {
-        payload.writeInt8(value instanceof BigDecimal ? ((BigDecimal) value).longValue() : (Long) value);
+        if (value instanceof BigDecimal) {
+            payload.writeInt8(((BigDecimal) value).longValue());
+        } else if (value instanceof Integer) {
+            payload.writeInt8(((Integer) value).longValue());
+        } else {
+            payload.writeInt8((Long) value);
+        }
     }
 }

